qiskit/circuit/qpy_serialization.py

```diff
@@ -673,6 +709,23 @@ def _read_parameter(file_obj):
     return param
 
 
+def _read_parameter_vec(file_obj, vectors):
+    param_raw = struct.unpack(PARAMETER_VECTOR_PACK, file_obj.read(PARAMETER_VECTOR_SIZE))
+    vec_name_size = param_raw[0]
+    param_uuid = uuid.UUID(bytes=param_raw[2])
+    param_index = param_raw[3]
+    name = file_obj.read(vec_name_size).decode("utf8")
+    if name not in vectors:
+        vectors[name] = ParameterVector(name, param_raw[1])
+    vector = vectors[name]
+    if vector[param_index]._uuid != param_uuid:
+        vector._params[param_index] = ParameterVectorElement.__new__(
+            ParameterVectorElement, vector, param_index, uuid=param_uuid
+        )
+        vector._params[param_index].__init__(vector, param_index)
+    return vector[param_index]
```

Comment:
The approach of constructing a new `ParameterVector` and then overwriting the `uuid` when we find a usage leaves the possibility of an unlikely but impossible to track down bug where `ParameterVectorElement`s would be `__eq__` or not across QPY serialization depending on their usage within the serialized circuit.

I'd expect this to be very unlikely, but is there any way we can guard against this? (One alternative might be always serializing the full `ParameterVector` the first time we see one of its `ParameterVectorElement`s, but I suspect that would be a larger change.)

Reply:
Well that's the tricky thing here, QPY doesn't have any context outside the `QuantumCircuit` object. If we wanted to serialize the full vectors we'd have to add a new global payload at the same level as the individual circuit (similar to the custom instructions table). It is possible but a larger change that changes the compatibility story for 0.19.1. Basically we go from having 0.19.1 be something that mostly can be loaded with 0.19.0 (which yes is explicitly unsupported, but it would work) with the exception of `ParameterVector`s and `PauliEvolutionGate` to something where the qpy generated with 0.19.1 is completely incompatible with 0.19.0's deserialization. I'm not sure if there is a practical use where this matters though we would be basically guarding against something like:

```
vec = ParameterVector(3)
qc = QuantumCircuit(1)
qc.rx(vec[1])
buf = io.BytesIO()
qpy_serialization.dump(qc, buf)
buf.seek(0)
new_circ = qpy_serialization.load(buf)
assert new_circ.parameters[0].vector[0] == vec[0]  # this will fail with the current branch
```

A simpler to implement fix is to just track in the `vectors` dict or similarly scoped structure which elements we've seen in the circuit and emit a warning if the vector list in the circuit is incomplete. Do you think this would be sufficient?

Reply:
I'd prefer to keep 0.19.1 with the small hack on top that we've implemented, and make larger changes in a subsequent release.

While compression isn't a huge goal of QPY, it seems slightly strange that we're reserialising all our parameters in every `ParameterExpression` - I'd have thought that all base resources of the circuit would be serialised up front in the header (in this case `Parameter` and `ParameterVector`), and then in `Instruction` handling, the `parameter` entry in the struct can just be an index (or other internal symbol-table lookup key), and `ParameterExpression` can be coded in a similar manner.  This would also be a way of solving Kevin's problem - we'd build the header up from `QuantumCircuit.parameters` up at the top, where we have all the information.

Reply:
Yeah I agree there are good reasons to do it longer term. I opened a tracking issue for this in 0.20.0: https://github.com/Qiskit/qiskit-terra/issues/7388 as part of QPY v4 (assuming there aren't anymore bugfix releases that require format revisions bumping this higher).

Reply:
I added a warning here: https://github.com/Qiskit/qiskit-terra/pull/7381/commits/14fa0060403821f79c288ebc3183ff0da68804a7 so if a `ParameterVector` isn't 100% identical via the UUIDs a `UserWarning` is emitted after circuit is finished being de-serialized. I opted for a warning because I feel like we should inform the user of this limitation but not block them from doing it. But, if others disagree we can make this an exception instead.

Reply:
> A simpler to implement fix is to just track in the vectors dict or similarly scoped structure which elements we've seen in the circuit and emit a warning if the vector list in the circuit is incomplete. Do you think this would be sufficient?

I this this is a good approach for heading this off in the immediate term, and should cover us until #7388 .